[WIP] Cross platform status bar API

janicduplessis opened the pull request:
I started working on improving the `StatusBar` API and make it work on Android. I added support for `setColor`, `setTranslucent` (the status bar is still visible but the app can draw under) and `setHidden` on Android. Looking for feedback on how to improve the API before I put more time on this :).

Right now I went for a cross platform API and functions that don't exist on a platform are just a no-op but I'm not sure it is the best choice since at the moment what is supported is very different between both platforms. I was wondering what you guys think and if it would be better off as 2 different modules.

It is also possible to port some of the features I added for Android to iOS even if there is no 'standard' way to do it. Like `setColor` could be implemented by drawing a colored view under the status bar and translucent by adding/removing some padding.

satya164 commented on ReactAndroid/src/main/java/com/facebook/react/modules/statusbar/StatusBarModule.java:
```diff
@@ -0,0 +1,120 @@
+  public void setColor(final int color, final boolean animated) {
+    final Activity activity = getCurrentActivity();
+    if (activity == null) {
+      throw new RuntimeException(ERROR_NO_ACTIVITY);
```
IMO it's be better to communicate to the JS code that it failed rather than crashing the app. I guess setColor can return a Promise.

janicduplessis replied:
@satya164 I went for an exception because I think changing the `StatusBar` when not attached to an activity is a programming mistake so it should fail fast. But now that unhandled promise rejections show a warning this accomplish the same thing without crashing in production so I agree this is a good change.

satya164 commented on Libraries/Components/StatusBar/StatusBar.js:
```diff
@@ -0,0 +1,76 @@
+
+  static setNetworkActivityIndicatorVisible(visible: boolean) {
+    if (Platform.OS === 'android') {
+      return;
```
Prolly `console.warn` that this is not available on Android?

satya164 commented:
Awesome. Happy to see more cross-platform APIs.

janicduplessis commented on Libraries/Components/StatusBar/StatusBar.js:
```diff
@@ -0,0 +1,76 @@
+  'light-content': string,
+}>;
+
+type StatusBarAnimation = $Enum<{
```
Should we add this to .eslintrc globals?

nicklockwood commented:
@janicduplessis I second @satya164's comment that it's great to see these APIs being ported.

In the case of StatusBarIOS, however, I've been thinking that it should really be a component rather than a module. I realized this when we got the request to add a `getColor` method.

The need for this makes sense, because a given component that wants to set the color when it's presented, and reset it when finished, has no way to know the original value. But it's a very un-React-like interaction.

It would be far nicer if you could embed a `<StatusBar>` component in your view, and it automatically takes effect when presented. Internally, RN would manage the StatusBar "stack", and push/pop the color and style attributes as the components are mounted and unmounted.

Does that sound like something you'd be interested in trying to implement?

janicduplessis commented:
@nicklockwood I really like the idea, I saw your comment about making it a component somewhere but wasn't sure how to implement that when there are multiple StatusBars in the view hierarchy. Poping/pusing when the component is mounted/unmounted sounds like a really good way to make this work and will work well with Navigator. I'll work on that!

nicklockwood commented:
@janicduplessis yeah, I'm not sure what to do if someone has multiple <StatusBar> components at sibling levels. We could just treat that as an error I suppose.

Another option would be to make the <StatusBar> a container, so you'd put your other views inside it, and it would always be the top level component in a tabbar, navigator, modal, etc. I'm not sure which approach is better TBH.

janicduplessis commented:
@nicklockwood I like the first option, I'll start with that and see if it seems good. We can just document the way it is supposed to be used.

satya164 commented:
I kinda prefer it as a container.

ide commented:
The thing about a container is that showing / hiding the status bar is tied to whether the container is mounted or unmounted. I don't think this is the right model when thinking about screens in a navigator -- rather than mounting, we care about whether the scene is visible or active.

janicduplessis commented:
Right now, showing/hiding the status bar is controlled by a prop on `StatusBar`. If no `StatusBar` component is added the app will just keep the default one or whatever is defined in the theme for android.

satya164 commented:
May be we should throw error if there are multiple statusbar components mounted at the same time?

janicduplessis commented:
Updated to be a component instead of a module. I am simply wrapping the existing iOS status bar module and the Android module I created with a component API by setting values when props change. This allows to implement the StatusBar stack logic once in JS.

I tested it with navigator in the UIExplorer app and it works like it should. You can use a top level StatusBar to set the default appearance then inside a route component you can add another StatusBar component and it will override the values that were set by the preceding one(s).

The only issue I hit is that since we are updating the values in the componentWillUnmount, the status bar animation is a little bit delayed compared to the navigator one.

cc @nicklockwood @satya164 @ide

nicklockwood commented:
Thanks! I'll take a look.

@facebook-github-bot import

satya164 commented:
@janicduplessis We should fix the flow error in `StatusBar.js` at least, if not the examples.

janicduplessis commented:
@satya164 I'm pretty sure they are all fixed. Some still show up in the diff comments because when I fixed them I didn't have to change those lines.

satya164 commented:
Oh, no probs. The bot is crazy.

nicklockwood commented:
@janicduplessis looking good so far. In the iOS example, the `animated` property doesn't seem to do anything - at least, setting it to false and then showing/hiding the bar doesn't disable the animation, but perhaps that's not its purpose?

janicduplessis commented:
@nicklockwood You are right, the animated value isn't getting used for hiding showing the bar. Actually, the status bar animation has 3 possible values 'none', 'slide' and 'fade'. Should we just keep animated and always use 'slide' or still support both maybe with another prop like animationType?

nicklockwood commented:
@janicduplessis maybe rename 'animated' to 'transition' and make it an enum of fade|slide|none?

janicduplessis commented:
I fixed the remaining TODOs, if this seems good to you it can be ready for review.

@nicklockwood I think I prefer to keep it as a simple animated prop since anyway the enum of fade|slide|none only makes sense for that one case. Maybe we could add a better way to control the animations for each prop in the future if it is needed but right now I like just specifying 'if it is animated or not' and not 'how to animate it'.

bestander commented:
@janicduplessis just a heads up.
Our CI now runs [BUCK](https://buckbuild.com/) https://circleci.com/gh/facebook/react-native/1852 in parallel with gradle.

We are working on making BUCK a first class build system for react native (as it is in our internal code base) but it will take a couple of weeks before we finish the migration and prepare the docs.

You would help us a lot (without it code won't be imported and we would have to patch the import manually) if you add a BUCK file to your new module, similar to [this](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/BUCK).

janicduplessis commented:
@bestander Thanks for the heads up, i'll make sure to test out building react native with buck and add the file :)

nicklockwood commented:
@facebook-github-bot shipit

nicklockwood commented:
(Sorry, meant to say import, not shipit - I need to do some testing internally first)

janicduplessis commented:
:+1:

nicklockwood commented:
@janicduplessis assuming I find no problems, is this ready to go as far as you're concerned?

janicduplessis commented:
@nicklockwood Yes I'm good with how it is now. I'll submit another PR after to add it to the website and improve the doc to explain better how it can be used.

nicklockwood commented on Libraries/Components/StatusBar/StatusBar.js:
```diff
@@ -0,0 +1,159 @@
+     * If the transition between status bar property changes should be animated.
+     * Supported for color, barStyle and hidden.
+     */
+    animated: React.PropTypes.bool,
```
While I mostly agree with your decision about keeping this as an "animated" prop, We'll need to add another prop to define the transition type for showing/hiding on iOS to avoid loss of functionality. Maybe:

    /*
     * The transition effect when showing and hiding the status bar. Defaults to 'fade'
     * @platform ios
     */
    showHideTransition: React.PropTypes.oneOf([
      'fade',
      'slide',
    ]),

janicduplessis replied:
what about `hiddenTransition` since the prop is called `hidden` it seems clearer to me.

nicklockwood replied:
I don't think that works. It makes it sound like "hidden" refers to the transition, not the other way around.

nicklockwood commented on ReactAndroid/src/main/java/com/facebook/react/modules/statusbar/StatusBarModule.java:
```diff
@@ -0,0 +1,129 @@
+
+  @Override
+  public String getName() {
+    return "StatusBarAndroid";
```
Why not call this RCTStatusBarManager? then you can avoid the forking in the JS code.

nicklockwood commented on Libraries/Components/StatusBar/StatusBar.js:
```diff
@@ -0,0 +1,159 @@
+const processColor = require('processColor');
+
+if (Platform.OS === 'ios') {
+  var RCTStatusBarManager = require('NativeModules').StatusBarManager;
```
Android module name should match iOS, then there's no need for this code fork.

nicklockwood commented on Examples/UIExplorer/StatusBarExample.js:
```diff
@@ -0,0 +1,181 @@
+            style={styles.wrapper}
+            onPress={() => this.setState({animated: !this.state.animated})}>
+            <View style={styles.button}>
+              <Text>Set animated `{!this.state.animated ? 'true' : 'false'}`</Text>
```
I found this example a bit confusing. When a button is labels "Set animated true", it's not entirely obvious that the current state is "false" and that the button will set it to "true" - instead it looks like maybe "true" is the current state.

I'd suggest making this unambiguous by changing the format of the button labels to:

    animated: true
    style: default
    etc...

That way the label reflects the current state.

nicklockwood commented on Libraries/Components/StatusBar/StatusBar.js:
```diff
@@ -0,0 +1,159 @@
+     * The color of the status bar.
+     * @platform android
+     */
+    color: ColorPropType,
```
Maybe rename this to `backgroundColor`, so that it's clear that it changes the color of the bar itself and not the text/icons?

nicklockwood commented on Libraries/Components/StatusBar/StatusBar.js:
```diff
@@ -0,0 +1,159 @@
+     *
+     * @platform android
+     */
+    translucent: React.PropTypes.bool,
```
Can this property be inferred from the backgroundColor? If it has an alpha component < 1.0 then it's translucent, otherwise it's inherently opaque.

janicduplessis commented:
@nicklockwood Thanks for the review! I made the changes you suggested, added some documentation about how the component can be used and added it to the website.

nicklockwood commented:
Sorry for the delay. Some internal tests got broken, and I'm just investigating. Nothing serious, I think.

marcshilling commented:
When I try to use this on Android, `StatusBarManager` is undefined. I'm on 0.20.0-rc1. Any ideas what's going on?
